main.go

```diff
@@ -125,84 +161,94 @@ func main() {
 	runtime.GOMAXPROCS(runtime.NumCPU())
 
 	app.Version(fmt.Sprintf("rev %s built with %s", buildRevision, buildCompiler))
+	app.Validate(validateFlags)
 	command := kingpin.MustParse(app.Parse(os.Args[1:]))
 
+	// metrics
+	if *graphiteAddr != nil {
+		logger.Printf("metrics enabled; reporting metrics via TCP to %s", *graphiteAddr)
+		go graphite.Graphite(metrics.DefaultRegistry, 1*time.Second, *metricsPrefix, *graphiteAddr)
+	}
+	if *metricsURL != "" {
+		logger.Printf("metrics enabled; reporting metrics via POST to %s", *metricsURL)
+	}
+	metrics := sqmetrics.NewMetrics(*metricsURL, *metricsPrefix, metrics.DefaultRegistry)
+
+	// wg used to gracefully exit on SIGTERM
+	listeners := &sync.WaitGroup{}
+	listeners.Add(1)
+
+	// Set up file watchers (if requested)
+	watcher := make(chan bool, 1)
+	if *timedReload > 0 {
+		go watchFiles([]string{*keystorePath, *caBundlePath}, time.Duration(*timedReload)*time.Second, watcher)
+	}
+
+	// channel to know if ghostunnel has started
+	started := make(chan bool, 1)
+
 	switch command {
 	case serverCommand.FullCommand():
+		if err := serverValidateFlags(); err != nil {
+			fmt.Fprintf(os.Stderr, "error: %s", err)
+			os.Exit(1)
+		}
 		logger.Printf("starting ghostunnel in server mode")
-		app.Validate(validateFlags)
 
-		err, dial := backendDialer()
+		err, dial := serverBackendDialer()
```

Comment:
No reason to change this in the same PR, but `serverBackendDialer` should probably return `err` as the second instead of the first argument